File: internal/flatpak/flatpak.go
// Package flatpak handles Flatpak package management operations.
package flatpak

import (
	"context"
	"strings"

	"github.com/iskry/cask/internal/executor"
)

// Install installs Flatpak packages.
func Install(ctx context.Context, exec executor.Executor, packages []string) (*executor.CommandResult, error) {
	if len(packages) == 0 {
		return &executor.CommandResult{Success: true, Stdout: "No Flatpak packages to install"}, nil
	}
	cmd := append([]string{"flatpak", "install", "-y", "--noninteractive"}, packages...)
	return exec.ExecuteSudo(ctx, cmd)
}

// Remove removes Flatpak packages.
func Remove(ctx context.Context, exec executor.Executor, packages []string) (*executor.CommandResult, error) {
	if len(packages) == 0 {
		return &executor.CommandResult{Success: true, Stdout: "No Flatpak packages to remove"}, nil
	}
	cmd := append([]string{"flatpak", "uninstall", "-y", "--noninteractive"}, packages...)
	return exec.ExecuteSudo(ctx, cmd)
}

// List returns installed Flatpak application IDs.
func List(ctx context.Context, exec executor.Executor) ([]string, error) {
	r, err := exec.Execute(ctx, []string{"flatpak", "list", "--app", "--columns=application"})
	if err != nil {
		return nil, err
	}
	if !r.Success {
		return nil, nil
	}
	var apps []string
	for _, line := range strings.Split(r.Stdout, "\n") {
		line = strings.TrimSpace(line)
		if line != "" {
			apps = append(apps, line)
		}
	}
	return apps, nil
}

// AddRemote adds a Flatpak remote repository.
func AddRemote(ctx context.Context, exec executor.Executor, name, url string) (*executor.CommandResult, error) {
	return exec.ExecuteSudo(ctx, []string{
		"flatpak", "remote-add", "--if-not-exists", name, url,
	})
}

File: src/cask/managers/flatpak.py
"""Flatpak operations."""
from __future__ import annotations

from cask.executor.protocol import Executor
from cask.result import Result


class FlatpakManager:
    """Manages Flatpak packages and remotes."""

    async def install(self, app_id: str, remote: str, exec: Executor) -> Result:
        r = await exec.execute_sudo(["flatpak", "install", "-y", remote, app_id])
        if r.exit_code == 0:
            return Result(ok=True, message=f"Installed {app_id}", actions=[f"flatpak install {app_id}"])
        return Result(ok=False, message=f"Failed to install {app_id}: {r.stderr}", actions=[])

    async def remove(self, app_id: str, exec: Executor) -> Result:
        r = await exec.execute_sudo(["flatpak", "uninstall", "-y", app_id])
        if r.exit_code == 0:
            return Result(ok=True, message=f"Removed {app_id}", actions=[f"flatpak uninstall {app_id}"])
        return Result(ok=False, message=f"Failed to remove {app_id}: {r.stderr}", actions=[])

    async def list_installed(self, exec: Executor) -> dict[str, str]:
        """Return {app_id: version} of installed apps."""
        r = await exec.execute(["flatpak", "list", "--app", "--columns=application,version"])
        apps = {}
        if r.exit_code == 0:
            for line in r.stdout.strip().splitlines():
                parts = line.split("\t")
                if len(parts) >= 2:
                    apps[parts[0]] = parts[1]
                elif len(parts) == 1 and parts[0]:
                    apps[parts[0]] = ""
        return apps

    async def add_remote(self, name: str, url: str, exec: Executor) -> Result:
        r = await exec.execute_sudo(["flatpak", "remote-add", "--if-not-exists", name, url])
        return Result(ok=r.exit_code == 0, message=f"Added remote {name}", actions=[])

    async def set_override(self, app_id: str, flags: list[str], exec: Executor) -> Result:
        r = await exec.execute_sudo(["flatpak", "override", "--system", app_id, *flags])
        return Result(ok=r.exit_code == 0, message=f"Set overrides for {app_id}", actions=flags)
